[Home](https://www.latex-project.org/)
[About](https://www.latex-project.org/about/)
[Get](https://www.latex-project.org/get/)
[LaTeX3](https://www.latex-project.org/latex3/)
[Publications](https://www.latex-project.org/publications/)
[Help](https://www.latex-project.org/help/)
[News](https://www.latex-project.org/news/)
***

![Latex Logo](https://www.latex-project.org/img/latex-project-logo.svg)

[Jupyter Math Reference](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Typesetting%20Equations.html)

Note: Some equations may not render correctly when viewing with Github's nbviewer. These equations are best when viewed directly served up by a Jupyter Notebook server.

### Latex Math Equations in a Jupyter Notebook
\begin{align}
f(x) = x^2
\end{align}

### The Lorenz Equations
\begin{align}
\dot{x} & = \sigma(y-x) \\
\dot{y} & = \rho x - y - xz \\
\dot{z} & = -\beta z + xy
\end{align}

### The Cauchy-Schwarz Inequality
\begin{equation*}
\left( \sum_{k=1}^n a_k b_k \right)^2 \leq \left( \sum_{k=1}^n a_k^2 \right) \left( \sum_{k=1}^n b_k^2 \right)
\end{equation*}

### A Cross Product Formula
\begin{equation*}
\mathbf{V}_1 \times \mathbf{V}_2 =  \begin{vmatrix}
\mathbf{i} & \mathbf{j} & \mathbf{k} \\
\frac{\partial X}{\partial u} &  \frac{\partial Y}{\partial u} & 0 \\
\frac{\partial X}{\partial v} &  \frac{\partial Y}{\partial v} & 0
\end{vmatrix}
\end{equation*}

### The probability of getting (k) heads when flipping (n) coins is
\begin{equation*}
P(E)   = {n \choose k} p^k (1-p)^{ n-k}
\end{equation*}

### An Identity of Ramanujan
\begin{equation*}
\frac{1}{\Bigl(\sqrt{\phi \sqrt{5}}-\phi\Bigr) e^{\frac25 \pi}} =
1+\frac{e^{-2\pi}} {1+\frac{e^{-4\pi}} {1+\frac{e^{-6\pi}}
{1+\frac{e^{-8\pi}} {1+\ldots} } } }
\end{equation*}

### A Rogers-Ramanujan Identity
\begin{equation*}
1 +  \frac{q^2}{(1-q)}+\frac{q^6}{(1-q)(1-q^2)}+\cdots =
\prod_{j=0}^{\infty}\frac{1}{(1-q^{5j+2})(1-q^{5j+3})},
\quad\quad \text{for $|q|<1$}.
\end{equation*}

### Maxwell’s Equations
\begin{align}
\nabla \times \vec{\mathbf{B}} -\, \frac1c\, \frac{\partial\vec{\mathbf{E}}}{\partial t} & = \frac{4\pi}{c}\vec{\mathbf{j}} \\   \nabla \cdot \vec{\mathbf{E}} & = 4 \pi \rho \\
\nabla \times \vec{\mathbf{E}}\, +\, \frac1c\, \frac{\partial\vec{\mathbf{B}}}{\partial t} & = \vec{\mathbf{0}} \\
\nabla \cdot \vec{\mathbf{B}} & = 0
\end{align}

### Sourcing
This expression $\sqrt{3x-1}+(1+x)^2$ is an example of a TeX inline equation in a [Markdown-formatted](https://daringfireball.net/projects/markdown/) sentence.

### Simple Greek Letter Equation
\begin{equation*}
\forall x \in X, \quad \exists y \leq \epsilon
\end{equation*}

### Greek Letters
\begin{equation*}
\alpha, A, \beta, B, \gamma, \Gamma, \pi, \Pi, \phi, \varphi, \mu, \Phi
\end{equation*}

### Limits
\begin{equation*}
\lim\limits_{x \to \infty} \exp(-x) = 0
\end{equation*}

### Modular arithmetic
\begin{equation*}
a \bmod b \\
a \pmod b
\end{equation*}

### Powers and indices
\begin{equation*}
k_{n+1} = n^2 + k_n^2 - k_{n-1} \\
\end{equation*}

For powers with more than one digit, surround the power with { }

\begin{equation*}
n^{22}\\
\end{equation*}

An underscore ( _ ) can be used with a vertical bar ( | ) to denote evaluation using subscript notation in mathematics

\begin{equation*}
f(n) = n^5 + 4n^2 + 2 |_{n=17}
\end{equation*}


### Aligning your equations
\begin{equation}
\begin{aligned}
F ={} & \{F_{x} \in  F_{c} : (|S| > |C|) \\
      & \cap (\mathrm{minPixels}  < |S| < \mathrm{maxPixels}) \\
      & \cap (|S_{\mathrm{conected}}| > |S| - \epsilon)\}
\end{aligned}
\end{equation}

### Fractions and Binomials
A fraction is created using the \frac{numerator}{denominator} command. Likewise, the binomial coefficient (aka the Choose function) may be written using the \binom command:

\begin{equation*}
\frac{n!}{k!(n-k)!} = \binom{n}{k}
\end{equation*}


You can embed fractions within fractions:

\begin{equation*}
\frac{\frac{1}{x}+\frac{1}{y}}{y-z}
\end{equation*}

### Modular arithmetic
\begin{equation*}
x^\frac{1}{2}
\end{equation*}

### Continued Fractions
\begin{equation}
  x = a_0 + \cfrac{1}{a_1 
          + \cfrac{1}{a_2 
          + \cfrac{1}{a_3 + \cfrac{1}{a_4} } } }
\end{equation}

### Multiplication of two numbers
To make multiplication visually similar to a fraction, a nested array can be used, for example multiplication of numbers written one below the other.

\begin{equation}
\frac{
    \begin{array}[b]{r}
      \left( x_1 x_2 \right)\\
      \times \left( x'_1 x'_2 \right)
    \end{array}
  }{
    \left( y_1y_2y_3y_4 \right)
  }
\end{equation}

### Roots
The \sqrt command creates a square root surrounding an expression.

\begin{equation}
\sqrt{\frac{a}{b}}
\end{equation}

It accepts an optional argument specified in square brackets ([ and ]) to change magnitude.

\begin{equation}
\sqrt[n]{1+x+x^2+x^3+\dots+x^n}
\end{equation}


### Sums and integrals

The \sum and \int commands insert the sum and integral symbols respectively, with limits specified using the caret (^) and underscore (_). The typical notation for sums is:

\begin{equation}
\sum_{i=1}^{10} t_i
\end{equation}

or

\begin{equation}
\displaystyle\sum_{i=1}^{10} t_i
\end{equation}

The limits for the integrals follow the same notation. It's also important to represent the integration variables with an upright d, which in math mode is obtained through the \mathrm{} command, and with a small space separating it from the integrand, which is attained with the \, command.

\begin{equation}
\int_0^\infty \mathrm{e}^{-x}\,\mathrm{d}x
\end{equation}

### Summation and Product

[Summation](https://en.wikipedia.org/wiki/Summation) is one of the most useful and commonly used symbols in iterative mathematics and data science. Despite it looking seemingly complex, it's quite simple and very useful.

Take for instance:

\begin{equation}
\sum_{i}^{n} t_i
\end{equation}

Where n is your limit or stopping number and i is your indexing start number (if i=1, you start at 1), you increment the value t until you reach n.

\begin{equation}
\sum_{i=1}^{5} t_i
\end{equation}

This equation simply states that we're going to sum the values from 1 to 5. 

All that this symbol represents is a for loop in the range of the number on top, starting from the number on the bottom. The variable set on the bottom becomes the index variable and any result per loop is added to an overall value. 

Here is the simple code for it below:

In [22]:
x = [1, 2, 3, 4, 5]
result = 0
for i in range(5):
    result += x[i]
print(result)

15


The Product Operator functions in the same manner, but instead of adding each result they will be multiplied.

\begin{equation}
\prod_{i}^{n} t_i
\end{equation}

\begin{equation}
\prod_{i=1}^{5} t_i
\end{equation}

In [23]:
x = [1, 2, 3, 4, 5]
result = 1
for i in range(5):
    result *= x[i]
print(result)

120


### Factorial

In mathematics, the [factorial](https://en.wikipedia.org/wiki/Factorial) of a positive integer n, denoted by n!, is the product of all positive integers less than or equal to n:

n! is the product of the natural numbers from 1 to n

- $n! = 1 x 2 x 3 x \dots xn $

for instance

- $5! = 1x2x3x4x5 = 120$

Negative numbers don't have a factorial and 0! = 1

- $n! = (n-1)!xn$
- $(n+1)! = n!$
- $(n+k)!$
- $(n-k)!$
- $(n+k)! = n!x(n+1)x(n+1)x\dots x(n+k)$

- $(n-k)! = \frac{n!}{(n-k+1)x(n-k+2)x\dots x(n-k+k)}$
- $(n-k)! = \frac{n!}{(n-k+1)x(n-k+2)x\dots xn}$

if n > k, then

- $ \frac{n!}{k!} = (k=1) x (k=2) x \dots xn$

so if $n = 7$ and $k = 4$, then

- $ \frac{7!}{4!} = 5 x 6 x 7 $

5! coded would look like:

In [24]:
result = 1
for i in range(1,6):
    result *= i
print(result)

120


### Brackets, braces and delimiters
The use of delimiters such as brackets soon becomes important when dealing with anything but the most trivial equations. Without them, formulas can become ambiguous. Also, special types of mathematical structures, such as matrices, typically rely on delimiters to enclose them.

There are a variety of delimiters available for use in LaTeX:
\begin{equation}
( a ), [ b ], \{ c \}, | d |, \| e \|,
\langle f \rangle, \lfloor g \rfloor,
\lceil h \rceil, \ulcorner i \urcorner
\end{equation}

where \lbrack and \rbrack may be used in place of [ and ].

### Automatic sizing
Very often mathematical features will differ in size, in which case the delimiters surrounding the expression should vary accordingly. This can be done automatically using the \left, \right, and \middle commands. Any of the previous delimiters may be used in combination with these:

\begin{equation}
\left(\frac{x^2}{y^3}\right)
\end{equation}

\begin{equation}
P\left(A=2\middle|\frac{A^2}{B}>4\right)
\end{equation}

Curly braces are defined differently by using \left\{ and \right\},

\begin{equation}
\left\{\frac{x^2}{y^3}\right\}
\end{equation}

If a delimiter on only one side of an expression is required, then an invisible delimiter on the other side may be denoted using a period (.).

\begin{equation}
\left.\frac{x^3}{3}\right|_0^1
\end{equation}

### Manual sizing
In certain cases, the sizing produced by the \left and \right commands may not be desirable, or you may simply want finer control over the delimiter sizes. In this case, the \big, \Big, \bigg and \Bigg modifier commands may be used:

\begin{equation}
( \big( \Big( \bigg( \Bigg(
\end{equation}

These commands are primarily useful when dealing with nested delimiters. For example, when typesetting

\begin{equation}
\frac{\mathrm d}{\mathrm d x} \left( k g(x) \right)
\end{equation}

we notice that the \left and \right commands produce the same size delimiters as those nested within it. This can be difficult to read. To fix this, we write

\begin{equation}
\frac{\mathrm d}{\mathrm d x} \big( k g(x) \big)
\end{equation}

Manual sizing can also be useful when an equation is too large, trails off the end of the page, and must be separated into two lines using an align command. Although the commands \left. and \right. can be used to balance the delimiters on each line, this may lead to wrong delimiter sizes. Furthermore, manual sizing can be used to avoid overly large delimiters if an \underbrace or a similar command appears between the delimiters.

### Matrices and arrays
A basic matrix may be created using the matrix environment: in common with other table-like structures, entries are specified by row, with columns separated using an ampersand (&) and a new rows separated with a double backslash (\\)

\begin{equation}
 \begin{matrix}
  a & b & c \\
  d & e & f \\
  g & h & i
 \end{matrix}
\end{equation}

When writing down arbitrary sized matrices, it is common to use horizontal, vertical and diagonal triplets of dots (known as ellipses) to fill in certain columns and rows. These can be specified using the \cdots, \vdots and \ddots respectively:

\begin{equation}
A_{m,n} = 
 \begin{pmatrix}
  a_{1,1} & a_{1,2} & \cdots & a_{1,n} \\
  a_{2,1} & a_{2,2} & \cdots & a_{2,n} \\
  \vdots  & \vdots  & \ddots & \vdots  \\
  a_{m,1} & a_{m,2} & \cdots & a_{m,n} 
 \end{pmatrix}
\end{equation}

In some cases you may want to have finer control of the alignment within each column, or want to insert lines between columns or rows. This can be achieved using the array environment, which is essentially a math-mode version of the tabular environment, which requires that the columns be pre-specified:

\begin{equation}
 \begin{array}{c|c}
  1 & 2 \\ 
  \hline
  3 & 4
 \end{array}
\end{equation}

You may see that the AMS matrix class of environments doesn't leave enough space when used together with fractions resulting in output similar to this:

\begin{equation}
 M = \begin{bmatrix}
       \frac{5}{6} & \frac{1}{6} & 0           \\[0.3em]
       \frac{5}{6} & 0           & \frac{1}{6} \\[0.3em]
       0           & \frac{5}{6} & \frac{1}{6}
     \end{bmatrix}
\end{equation}

### Matrices in running text
To insert a small matrix, and not increase leading in the line containing it, use smallmatrix environment:

A matrix in text must be set smaller:
$\bigl(\begin{smallmatrix}
a&b \\ c&d
\end{smallmatrix} \bigr)$
to not increase leading in a portion of text.

### Adding text to equations
The math environment differs from the text environment in the representation of text. Here is an example of trying to represent text within the math environment:
\begin{equation}
50 apples \times 100 apples = lots of apples^2
\end{equation}

There are two noticeable problems: there are no spaces between words or numbers, and the letters are italicized and more spaced out than normal. Both issues are simply artifacts of the maths mode, in that it treats it as a mathematical expression: spaces are ignored (LaTeX spaces mathematics according to its own rules), and each character is a separate element (so are not positioned as closely as normal text).

There are a number of ways that text can be added properly. The typical way is to wrap the text with the \text{...} command (a similar command is \mbox{...}, though this causes problems with subscripts, and has a less descriptive name). Let's see what happens when the above equation code is adapted:

\begin{equation}
50 \text{apples} \times 100 \text{apples} = \text{lots of apples}^2
\end{equation}

The text looks better. However, there are no gaps between the numbers and the words. Unfortunately, you are required to explicitly add these. There are many ways to add spaces between maths elements, but for the sake of simplicity we may simply insert space characters into the \text commands.
 
\begin{equation}
50 \text{ apples} \times 100 \text{ apples} = \text{lots of apples}^2
\end{equation}

### Formatted text
Using the \text is fine and gets the basic result. Yet, there is an alternative that offers a little more flexibility. You may recall the introduction of font formatting commands, such as \textrm, \textit, \textbf, etc. These commands format the argument accordingly, e.g., \textbf{bold text} gives bold text. These commands are equally valid within a maths environment to include text. The added benefit here is that you can have better control over the font formatting, rather than the standard text achieved with \text.
\begin{equation}
50 \textrm{ apples} \times 100
 \textbf{ apples} = \textit{lots of apples}^2
\end{equation}

### Formatting mathematics symbols

These formatting commands can be wrapped around the entire equation, and not just on the textual elements: they only format letters, numbers, and uppercase Greek, and other math commands are unaffected.

To bold lowercase Greek or other symbols use the \boldsymbol command; this will only work if there exists a bold version of the symbol in the current font. As a last resort there is the \pmb command (poor man's bold): this prints multiple versions of the character slightly offset against each other.

\begin{equation}
\boldsymbol{\beta} = (\beta_1,\beta_2,\dotsc,\beta_n)
\end{equation}

To change the size of the fonts in math mode, see Changing font size.

### Accents
So what to do when you run out of symbols and fonts? Well, the next step is to use accents:

\begin{equation}
a'  a'' \hat{a} \bar{a} \grave{a} \acute{a} \dot{a}	\ddot{a} \not{a} \mathring{a}	
\overrightarrow{AB}	 \overleftarrow{AB}	a''' a''''	\overline{aaa}	\check{a} \breve{a}	\vec{a}	\dddot{a}		\ddddot{a}	\widehat{AAA} \widetilde{AAA} \stackrel\frown{AAA}	\tilde{a}	\underline{a} 
\end{equation}

### Color
The package xcolor, described in Colors, allows us to add color to our equations. 

\begin{equation}
k = {\color{red}x} \mathbin{\color{blue}-} 2
\end{equation}

### Plus and minus signs

LaTeX deals with the + and − signs in two possible ways. The most common is as a binary operator. When two maths elements appear on either side of the sign, it is assumed to be a binary operator, and as such, allocates some space to either side of the sign. The alternative way is a sign designation. This is when you state whether a mathematical quantity is either positive or negative. This is common for the latter, as in math, such elements are assumed to be positive unless a − is prefixed to it. In this instance, you want the sign to appear close to the appropriate element to show their association. If you put a + or a − with nothing before it but you want it to be handled like a binary operator you can add an invisible character before the operator using {}. This can be useful if you are writing multiple-line formulas, and a new line could start with a − or +, for example, then you can fix some strange alignments adding the invisible character where necessary.

A plus-minus sign is written as:
\begin{equation}
\pm
\end{equation}


Similarly, there exists also a minus-plus sign:

\begin{equation}
\mp
\end{equation}

### Controlling horizontal spacing
LaTeX is obviously pretty good at typesetting maths—it was one of the chief aims of the core TeX system that LaTeX extends. However, it can't always be relied upon to accurately interpret formulas in the way you did. It has to make certain assumptions when there are ambiguous expressions. The result tends to be slightly incorrect horizontal spacing. In these events, the output is still satisfactory, yet any perfectionists will no doubt wish to fine-tune their formulas to ensure spacing is correct. These are generally very subtle adjustments.

There are other occasions where LaTeX has done its job correctly, but you just want to add some space, maybe to add a comment of some kind. For example, in the following equation, it is preferable to ensure there is a decent amount of space between the maths and the text.

\begin{equation}
 f(n) =
  \begin{cases}
    n/2       & \quad \text{if } n \text{ is even}\\
    -(n+1)/2  & \quad \text{if } n \text{ is odd}
  \end{cases}
\end{equation}


OK, so back to the fine tuning as mentioned at the beginning of the document. A good example would be displaying the simple equation for the indefinite integral of y with respect to x:
if you were to try this, you may write:

\begin{equation}
\int y \mathrm{d}x
\end{equation}

However, this doesn't give the correct result. LaTeX doesn't respect the white-space left in the code to signify that the y and the dx are independent entities. Instead, it lumps them altogether. A \quad would clearly be overkill in this situation—what is needed are some small spaces to be utilized in this type of instance, and that's what LaTeX provides:

\begin{equation}
\int y\, \mathrm{d}x
\end{equation}

\begin{equation}
\int y\: \mathrm{d}x
\end{equation}

\begin{equation}
\int y\; \mathrm{d}x
\end{equation}

The negative space may seem like an odd thing to use, however, it wouldn't be there if it didn't have some use! Take the following example:

\begin{equation}
\left(
    \begin{array}{c}
      n \\
      r
    \end{array}
  \right) = \frac{n!}{r!(n-r)!}
\end{equation}

The matrix-like expression for representing binomial coefficients is too padded. There is too much space between the brackets and the actual contents within. This can easily be corrected by adding a few negative spaces after the left bracket and before the right bracket.

\begin{equation}
\left(\!
    \begin{array}{c}
      n \\
      r
    \end{array}
  \!\right) = \frac{n!}{r!(n-r)!}
\end{equation}

n any case, adding some spaces manually should be avoided whenever possible: it makes the source code more complex and it's against the basic principles of a What You See is What You Mean approach. The best thing to do is to define some commands using all the spaces you want and then, when you use your command, you don't have to add any other space. Later, if you change your mind about the length of the horizontal space, you can easily change it modifying only the command you defined before. Let us use an example: you want the d of a dx in an integral to be in roman font and a small space away from the rest. If you want to type an integral like \int x \, \mathrm{d} x, you can define a command like this:


\newcommand{\dd}{\mathop{}\,\mathrm{d}}


in the preamble of your document. We have chosen \dd just because it reminds the "d" it replaces and it is fast to type. Doing so, the code for your integral becomes \int x \dd x. Now, whenever you write an integral, you just have to use the \dd instead of the "d", and all your integrals will have the same style. If you change your mind, you just have to change the definition in the preamble, and all your integrals will be changed accordingly.

[Source](https://en.wikibooks.org/wiki/LaTeX/Mathematics)

## Mean, Median, Mode

Mean, median, and mode are different measures of center in a numerical data set. They each try to summarize a dataset with a single number to represent a "typical" data point from the dataset.

__Mean__ The "average" number found by adding all data points and dividing by the number of data points.

To get the mean

\begin{equation}
\bar{y} = \frac{\sum\nolimits y}{N}
\end{equation}

- $\sum\nolimits y$ Take all the values and add them up
- divide by the total number of observations you have


__Median__ The middle number found by ordering all data points and picking out the one in the middle (or if there are two middle numbers, taking the mean of those two numbers).


To get the median

\begin{equation}
y = \tilde{x}
\end{equation}

$ y = [1,2,3,4,4,3,2,1] $

$ y = 4 $


The __mode__ is the most commonly occurring data point in a dataset. The mode is useful when there are a lot of repeated values in a dataset. There can be no mode, one mode, or multiple modes in a dataset.

## Standard Deviation

__Standard deviation__ is a number used to tell how measurements for a group are spread out from the average (mean), or expected value. A low standard deviation means that most of the numbers are close to the average. A high standard deviation means that the numbers are more spread out.

$s = \sqrt{\frac{1}{N-1} \sum_{i=1}^N (x_i - \overline{x})^2}$


## Variance

The __Variance__ is defined as: The average of the squared differences from the Mean. To calculate the variance follow these steps: Find the __Mean__, Then for each number: subtract the __Mean__ and square the result (the squared difference).

$\sigma^2 = \frac{\displaystyle\sum_{i=1}^{n}(x_i - \mu)^2} {n}$

## What is RMSE

 $RMSE = \sqrt{\frac{1}{n}\Sigma_{i=1}^{n}{\Big(\frac{d_i -f_i}{\sigma_i}\Big)^2}}$
 
 
 ## Root mean squared error (RMSE)
RMSE is a quadratic scoring rule that also measures the average magnitude of the error. It’s the square root of the average squared differences between prediction and actual observation. 

 $RMSE = \sqrt{\frac{1}{n}\Sigma_{i=1}^{n}{\Big(\frac{d_i -f_i}{\sigma_i}\Big)^2}}$
 
 we get this easy from sklearn:

```python
from sklearn.metrics import mean_squared_error
from math import sqrt
rmse = sqrt(mean_squared_error(y_actual, y_predicted))
print(rmse)

# another way
n = len(predictions)
rmse = np.linalg.norm(predictions - targets) / np.sqrt(n)

# and another way
from sklearn.metrics import mean_squared_error
from math import sqrt
expected = [0.0, 0.5, 0.0, 0.5, 0.0]
predictions = [0.2, 0.4, 0.1, 0.6, 0.2]
mse = mean_squared_error(expected, predictions)
rmse = sqrt(mse)
print('RMSE: %f' % rmse)

```
The mean square root and square root of it will be useful.


In [25]:
from sklearn.metrics import mean_squared_error
from math import sqrt
expected = [0.0, 0.5, 0.0, 0.5, 0.0]
predictions = [0.2, 0.4, 0.1, 0.6, 0.2]
mse = mean_squared_error(expected, predictions)
rmse = sqrt(mse)
print('RMSE: %f' % rmse)

RMSE: 0.148324
